staging/src/k8s.io/kube-apiextensions-server/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -15,3 +15,137 @@ limitations under the License.
+		allErrs = append(allErrs, field.Required(fldPath.Child("group"), ""))
+	}
+	if errs := validationutil.IsDNS1123Subdomain(spec.Group); len(errs) > 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("group"), spec.Group, strings.Join(errs, ",")))
+	}
+	if len(strings.Split(spec.Group, ".")) < 2 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("group"), spec.Group, "should be a domain with at least one dot"))
+	}
+
+	if len(spec.Version) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("version"), ""))
+	}
+	if errs := validationutil.IsDNS1035Label(spec.Version); len(errs) > 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("version"), spec.Version, strings.Join(errs, ",")))
+	}
+
+	switch spec.Scope {
+	case apiextensions.ClusterScoped, apiextensions.NamespaceScoped:
+	default:
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("scope"), spec.Scope, []string{string(apiextensions.ClusterScoped), string(apiextensions.NamespaceScoped)}))
+	}
+
+	// in addition to the basic name restrictions, some names are required for spec, but not for status
+	if len(spec.Names.Plural) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("names", "plural"), ""))
+	}
+	if len(spec.Names.Singular) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("names", "singular"), ""))
+	}
+	if len(spec.Names.Kind) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("names", "kind"), ""))
+	}
+	if len(spec.Names.ListKind) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("names", "listKind"), ""))
+	}
+
+	allErrs = append(allErrs, ValidateCustomResourceNames(&spec.Names, fldPath.Child("names"))...)
+
+	return allErrs
+}
+
+func ValidateCustomResourceSpecUpdate(spec, oldSpec *apiextensions.CustomResourceSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := ValidateCustomResourceSpec(spec, fldPath)
+
+	// these all affect the storage, so you can't change them
+	genericvalidation.ValidateImmutableField(spec.Group, oldSpec.Group, fldPath.Child("group"))
+	genericvalidation.ValidateImmutableField(spec.Version, oldSpec.Version, fldPath.Child("version"))
+	genericvalidation.ValidateImmutableField(spec.Scope, oldSpec.Scope, fldPath.Child("scope"))
+	genericvalidation.ValidateImmutableField(spec.Names.Kind, oldSpec.Names.Kind, fldPath.Child("names", "kind"))
+
+	// this affects the expected resource name, so you can't change it either
+	genericvalidation.ValidateImmutableField(spec.Names.Plural, oldSpec.Names.Plural, fldPath.Child("names", "plural"))
+
+	return allErrs
+}
+func ValidateCustomResourceStatus(status *apiextensions.CustomResourceStatus, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateCustomResourceNames(&status.AcceptedNames, fldPath.Child("acceptedNames"))...)
+	return allErrs
+}
+
+func ValidateCustomResourceNames(names *apiextensions.CustomResourceNames, fldPath *field.Path) field.ErrorList {
```

Comment:
> Do you want to allow empty Kind and Plural? Plural could be derived from the CustomResource name, but Kind?

This used by status validation too.  Status may not have all these field set.  If they are set, they must be valid, but if they are not set how to handle that varies with where the field is embedded.  See the spec validation above.

Also, no information is derived post-defaulting: https://github.com/kubernetes/kubernetes/pull/45561